File: Behavioural-Patterns/Command/C#/Command/Installer.cs
public class Installer {
    private readonly List<ICommand> commands = new List<ICommand>();
    private readonly Stack<ICommand> commandsHistory = new Stack<ICommand>();

    public void addCommand(ICommand command){
        commands.Add(command);
    }

    public void install(){
        Console.WriteLine("--- Starting Installation ---");
        try {
            foreach (ICommand command in commands) {
                command.execute();
                commandsHistory.Push(command);
            }
            Console.WriteLine("\n--- Installation Completed Successfully! ---");
        } catch (Exception e) {
            Console.WriteLine("\n*** ERROR: Installation failed. Reason: " + e.Message + " ***");
            rollback();
        }
    }

    private void rollback() {
        Console.WriteLine("\n--- Starting Rollback ---");
        while (commandsHistory.Count > 0) {
            ICommand lastCommand = commandsHistory.Pop();
            lastCommand.undo();
        }
        Console.WriteLine("--- Rollback Complete. System is in its original state. ---");
    }  
}

File: Behavioural-Patterns/Command/Java/Installer.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

import Commands.ICommand;

public class Installer {
    private final List<ICommand> commands = new ArrayList<>();
    private final Deque<ICommand> commandsHistory = new ArrayDeque<>();

    public void addCommand(ICommand command){
        commands.add(command);
    }

    public void install(){
        System.out.println("--- Starting Installation ---");
        try {
            for (ICommand command : commands) {
                command.execute();
                commandsHistory.push(command);
            }
            System.out.println("\n--- Installation Completed Successfully! ---");
        } catch (Exception e) {
            System.err.println("\n*** ERROR: Installation failed. Reason: " + e.getMessage() + " ***");
            rollback();
        }
    }

    private void rollback() {
        System.out.println("\n--- Starting Rollback ---");
        while (!commandsHistory.isEmpty()) {
            ICommand lastCommand = commandsHistory.pop();
            lastCommand.undo();
        }
        System.out.println("--- Rollback Complete. System is in its original state. ---");
    }  
}
